### PR TITLE
Physics Update 11

### PR DESCRIPTION
**Pull request by CedricGuillemet**

### CODE REVIEW

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

File: packages/dev/core/src/Physics/v2/physicsShape.ts [hunk 1]
```diff
@@ -223,13 +249,27 @@ export class PhysicsShapeCylinder extends PhysicsShape {
...
+        return new PhysicsShapeCylinder(pointA, pointB, radius, mesh.getScene());
+    }
 }
 
 /**
  * Helper object to create a box shape
  */
 export class PhysicsShapeBox extends PhysicsShape {
-    /**
+    /**getHierarchyBoundingVectors
```
Comment (RaananW) [inline packages/dev/core/src/Physics/v2/physicsShape.ts:272]: Is this change intentional?

File: packages/dev/core/src/Physics/v2/physicsBody.ts [hunk 2]
```diff
@@ -338,14 +375,18 @@ export class PhysicsBody {
      * Filtering by body is inefficient. It's more preferable to register a collision callback for the entire world
      * and do the filtering on the user side.
      */
-    public registerOnCollide(func: (collider: PhysicsBody, collidedAgainst: PhysicsBody, point: Nullable<Vector3>) => void): void {
+    public registerOnCollide(
+        func: (collider: PhysicsBody, collidedAgainst: PhysicsBody, point: Nullable<Vector3>, distance: number, impulse: number, normal: Nullable<Vector3>) => void
+    ): void {
```
[the remark this answers is not in the document]
Comment (CedricGuillemet) [reply packages/dev/core/src/Physics/v2/physicsBody.ts:380]: Makes sense. Same for body collision callback registration?

[the remark this answers is not in the document]
Comment (CedricGuillemet) [reply packages/dev/core/src/Physics/v2/physicsBody.ts:380]: I'll do the change in another PR as I need to test with a bunch of PG in another repo...

File: packages/dev/core/src/Physics/v2/physicsShape.ts [hunk 3]
```diff
@@ -223,13 +249,27 @@ export class PhysicsShapeCylinder extends PhysicsShape {
     constructor(pointA: Vector3, pointB: Vector3, radius: number, scene: Scene) {
         super({ type: ShapeType.CYLINDER, parameters: { pointA: pointA, pointB: pointB, radius: radius } }, scene);
     }
+
+    /**
+     * Derive an approximate cylinder from the transform node. Note, this is
+     * not the optimal bounding cylinder.
+     * @param TransformNode node Node from which to derive a cylinder shape
+     */
+    static fromMesh(mesh: AbstractMesh): PhysicsShapeCylinder {
```
[the remark this answers is not in the document]
Comment (CedricGuillemet) [reply packages/dev/core/src/Physics/v2/physicsShape.ts]: done.

File: packages/dev/core/src/Physics/v2/physicsShape.ts [hunk 4]
```diff
@@ -207,6 +219,20 @@ export class PhysicsShapeCapsule extends PhysicsShape {
     constructor(pointA: Vector3, pointB: Vector3, radius: number, scene: Scene) {
         super({ type: ShapeType.CAPSULE, parameters: { pointA: pointA, pointB: pointB, radius: radius } }, scene);
     }
+
+    /**
+     * Derive an approximate capsule from the transform node. Note, this is
+     * not the optimal bounding capsule.
+     * @param TransformNode node Node from which to derive a cylinder shape
+     */
+    static fromMesh(mesh: AbstractMesh): PhysicsShapeCapsule {
```
[the remark this answers is not in the document]
Comment (CedricGuillemet) [reply packages/dev/core/src/Physics/v2/physicsShape.ts]: done.

File: packages/dev/core/src/Physics/v2/physicsBody.ts [hunk 5]
```diff
@@ -338,14 +375,18 @@ export class PhysicsBody {
...
         return this._physicsPlugin.registerOnBodyCollide(this, func);
     }
 
     /**
      * Unregister a collision callback that is called when the body collides
      */
-    public unregisterOnCollide(func: (collider: PhysicsBody, collidedAgainst: PhysicsBody, point: Nullable<Vector3>) => void): void {
+    public unregisterOnCollide(
+        func: (collider: PhysicsBody, collidedAgainst: PhysicsBody, point: Nullable<Vector3>, distance: number, impulse: number, normal: Nullable<Vector3>) => void
+    ): void {
```
[the remark this answers is not in the document]
Comment (CedricGuillemet) [reply packages/dev/core/src/Physics/v2/physicsBody.ts:389]: yes, same as above. will do another PR for that.

File: packages/dev/core/src/Physics/v2/physicsAggregate.ts [hunk 6]
```diff
@@ -158,15 +169,22 @@ export class PhysicsAggregate {
         this._options.friction = _options.friction === void 0 ? 0.2 : _options.friction;
         this._options.restitution = _options.restitution === void 0 ? 0.2 : _options.restitution;
 
-        this.body = new PhysicsBody(transformNode, this._scene);
+        const motionType = this._options.mass == 0 ? PhysicsMotionType.STATIC : PhysicsMotionType.DYNAMIC;
```
[the remark this answers is not in the document]
Comment (CedricGuillemet) [reply packages/dev/core/src/Physics/v2/physicsAggregate.ts]: fixed

File: packages/dev/core/src/Physics/v2/IPhysicsEnginePlugin.ts [hunk 7]
```diff
@@ -59,23 +59,65 @@ export enum ConstraintMotorType {
 
 /** @internal */
 export interface PhysicsShapeParameters {
+    /**
+     *
+     */
```
[the remark this answers is not in the document]
Comment (CedricGuillemet) [reply packages/dev/core/src/Physics/v2/IPhysicsEnginePlugin.ts]: documented

File: packages/dev/core/src/Physics/v2/physicsBody.ts [hunk 8]
```diff
@@ -173,6 +188,28 @@ export class PhysicsBody {
...
+    }
+
+    public getMotionType(): PhysicsMotionType {
+        return this._physicsPlugin.getMotionType(this);
+    }
+
+    /**
+     * Computes the mass properties of the physics object, based on the set of physics shapes this body uses.
+     *
+     * @param massProps - The mass properties to set.
```
Comment (sebavan) [inline packages/dev/core/src/Physics/v2/physicsBody.ts:202]: There is no param :-)

File: packages/dev/core/src/Physics/v2/physicsBody.ts [hunk 9]
```diff
@@ -338,14 +375,18 @@ export class PhysicsBody {
      * Filtering by body is inefficient. It's more preferable to register a collision callback for the entire world
      * and do the filtering on the user side.
      */
-    public registerOnCollide(func: (collider: PhysicsBody, collidedAgainst: PhysicsBody, point: Nullable<Vector3>) => void): void {
+    public registerOnCollide(
```
Comment (sebavan) [inline packages/dev/core/src/Physics/v2/physicsBody.ts:378]: params are missing in the doc. @RaananW  I wonder why the doc check did not caught them ?

File: packages/dev/core/src/Physics/v2/physicsShape.ts
[the remark this answers is not in the document]
Comment (CedricGuillemet) [reply packages/dev/core/src/Physics/v2/physicsShape.ts:1]: I take a note on that. There might be other places to change. will not be part on this PR.

File: packages/dev/core/src/Physics/v2/physicsBody.ts [hunk 10]
```diff
@@ -338,14 +375,18 @@ export class PhysicsBody {
...
+        func: (collider: PhysicsBody, collidedAgainst: PhysicsBody, point: Nullable<Vector3>, distance: number, impulse: number, normal: Nullable<Vector3>) => void
+    ): void {
         return this._physicsPlugin.registerOnBodyCollide(this, func);
     }
 
     /**
      * Unregister a collision callback that is called when the body collides
      */
-    public unregisterOnCollide(func: (collider: PhysicsBody, collidedAgainst: PhysicsBody, point: Nullable<Vector3>) => void): void {
+    public unregisterOnCollide(
```
Comment (sebavan) [inline packages/dev/core/src/Physics/v2/physicsBody.ts:387]: Doc is missing for the param as well

**Review by sebavan — COMMENTED** [review]
Overall LGTM but I am super noob with physics :-)

It seems that functions/method docs is sometimes not correct cc @RaananW for this part in case we have an issue there

File: packages/dev/core/src/Physics/v2/physicsBody.ts (hunk 8 above)
Comment (CedricGuillemet) [reply packages/dev/core/src/Physics/v2/physicsBody.ts:202]: this code has been discussed in this PR. see previous comments.

File: packages/dev/core/src/Physics/v2/physicsBody.ts (hunk 9 above)
Comment (CedricGuillemet) [reply packages/dev/core/src/Physics/v2/physicsBody.ts:378]: this code has been discussed in this PR. see previous comments.

File: packages/dev/core/src/Physics/v2/physicsBody.ts (hunk 10 above)
Comment (CedricGuillemet) [reply packages/dev/core/src/Physics/v2/physicsBody.ts:387]: this code has been discussed in this PR. see previous comments.

File: packages/dev/core/src/Physics/v2/physicsBody.ts (hunk 8 above)
Comment (CedricGuillemet) [reply packages/dev/core/src/Physics/v2/physicsBody.ts:202]: Fixed

**Review by RaananW — APPROVED** [review]
LGTM, reminder (for me as well :-)) about the discussions of callback registration and getters/setters.
